python/ffsim/tenpy/gates/basic_gates.py

```diff
@@ -0,0 +1,261 @@
+
+    # define total gate
+    if spin is Spin.ALPHA:
+        Ggate = Ggate_a
+    elif spin is Spin.BETA:
+        Ggate = Ggate_b
+    elif spin is Spin.ALPHA_AND_BETA:
+        Ggate = Ggate_a @ Ggate_b
+    else:
+        raise ValueError("undefined spin")
+
+    # convert to (N, Sz)-symmetry-conserved basis
+    Ggate_sym = _sym_cons_basis(Ggate)
+
+    return Ggate_sym
+
+
+def num_interaction(theta: float, spin: Spin = Spin.ALPHA_AND_BETA) -> np.ndarray:
+    r"""The number interaction gate.
+
+    The number interaction gate as defined in
+    `apply_num_interaction <https://qiskit-community.github.io/ffsim/api/ffsim.html#ffsim.apply_num_interaction>`__,
+    returned in the TeNPy (N, Sz)-symmetry-conserved basis.
+
+    Args:
+        theta: The rotation angle.
+        spin: Choice of spin sector(s) to act on.
+
+            - To act on only spin alpha, pass :const:`ffsim.Spin.ALPHA`.
+            - To act on only spin beta, pass :const:`ffsim.Spin.BETA`.
+            - To act on both spin alpha and spin beta, pass
+              :const:`ffsim.Spin.ALPHA_AND_BETA` (this is the default value).
+
+    Returns:
+        The number interaction gate in the TeNPy (N, Sz)-symmetry-conserved basis.
+    """
+
+    # alpha sector / up spins
+    if spin in [Spin.ALPHA, Spin.ALPHA_AND_BETA]:
+        # # using TeNPy SpinHalfFermionSite(cons_N=None, cons_Sz=None) operators
+        # Ngate_a = sp.linalg.expm(1j * theta * Nu)
+        Ngate_a = np.eye(4, dtype=complex)
+        for i in [1, 3]:
+            Ngate_a[i, i] = cmath.exp(1j * theta)
+
+    # beta sector / down spins
+    if spin in [Spin.BETA, Spin.ALPHA_AND_BETA]:
+        # # using TeNPy SpinHalfFermionSite(cons_N=None, cons_Sz=None) operators
+        # Ngate_b = sp.linalg.expm(1j * theta * Nd)
+        Ngate_b = np.eye(4, dtype=complex)
+        for i in [2, 3]:
+            Ngate_b[i, i] = cmath.exp(1j * theta)
+
+    # define total gate
+    if spin is Spin.ALPHA:
+        Ngate = Ngate_a
+    elif spin is Spin.BETA:
+        Ngate = Ngate_b
+    elif spin is Spin.ALPHA_AND_BETA:
+        Ngate = Ngate_a @ Ngate_b
+    else:
+        raise ValueError("undefined spin")
+
+    # convert to (N, Sz)-symmetry-conserved basis
+    Ngate_sym = _sym_cons_basis(Ngate)
+
+    return Ngate_sym
+
+
+def on_site_interaction(theta: float) -> np.ndarray:
+    r"""The on-site interaction gate.
+
+    The on-site interaction gate as defined in
+    `apply_on_site_interaction <https://qiskit-community.github.io/ffsim/api/ffsim.html#ffsim.apply_on_site_interaction>`__,
+    returned in the TeNPy (N, Sz)-symmetry-conserved basis.
+
+    Args:
+        theta: The rotation angle.
+
+    Returns:
+        The on-site interaction gate in the TeNPy (N, Sz)-symmetry-conserved basis.
+    """
+
+    # # using TeNPy SpinHalfFermionSite(cons_N=None, cons_Sz=None) operators
+    # OSgate = sp.linalg.expm(1j * theta * Nu @ Nd)
+    OSgate = np.eye(4, dtype=complex)
+    OSgate[3, 3] = cmath.exp(1j * theta)
```

Comment:
```suggestion
    OSgate = np.diag([1, 1, 1, cmath.exp(1j * theta)])
```

Similarly for other places where you construct diagonal matrices.